metagpt/provider/unify.py

```diff
@@ -0,0 +1,122 @@
+from openai.types.chat import ChatCompletion
+
+from metagpt.configs.llm_config import LLMConfig, LLMType
+from metagpt.const import USE_CONFIG_TIMEOUT
+from metagpt.logs import log_llm_stream, logger
+from metagpt.provider.base_llm import BaseLLM
+from metagpt.provider.llm_provider_registry import register_provider
+from metagpt.utils.cost_manager import CostManager
+from metagpt.utils.token_counter import count_message_tokens, OPENAI_TOKEN_COSTS
+from unify.clients import Unify, AsyncUnify
+
+@register_provider([LLMType.UNIFY])
+class UnifyLLM(BaseLLM):
+    def __init__(self, config: LLMConfig):
+        self.config = config
+        self._init_client()
+        self.cost_manager = CostManager(token_costs=OPENAI_TOKEN_COSTS)  # Using OpenAI costs as Unify is compatible
+
+    def _init_client(self):
+        self.model = self.config.model
+        self.client = Unify(
+            api_key=self.config.api_key,
+            endpoint=f"{self.config.model}@{self.config.provider}",
+        )
+        self.async_client = AsyncUnify(
+            api_key=self.config.api_key,
+            endpoint=f"{self.config.model}@{self.config.provider}",
+        )
+
+    def _const_kwargs(self, messages: list[dict], stream: bool = False) -> dict:
+        return {
+            "messages": messages,
+            "max_tokens": self.config.max_token,
+            "temperature": self.config.temperature,
+            "stream": stream,
+        }
+
+    def get_choice_text(self, resp: Union[ChatCompletion, str]) -> str:
+        if isinstance(resp, str):
+            return resp
+        return resp.choices[0].message.content if resp.choices else ""
+
+    def _update_costs(self, usage: dict):
+        prompt_tokens = usage.get("prompt_tokens", 0)
+        completion_tokens = usage.get("completion_tokens", 0)
+        self.cost_manager.update_cost(prompt_tokens, completion_tokens, self.model)
+
+    async def _achat_completion(self, messages: list[dict], timeout=USE_CONFIG_TIMEOUT) -> ChatCompletion:
+        try:
+            response = await self.async_client.generate(
+                messages=messages,
+                max_tokens=self.config.max_token,
+                temperature=self.config.temperature,
+                stream=False,
+            )
+            # Construct a ChatCompletion object to match OpenAI's format
+            chat_completion = ChatCompletion(
+                id="unify_chat_completion",
+                object="chat.completion",
+                created=0,  # Unify doesn't provide this, so we use 0
+                model=self.model,
+                choices=[{
+                    "index": 0,
+                    "message": {
+                        "role": "assistant",
+                        "content": response,
+                    },
+                    "finish_reason": "stop",
+                }],
+                usage=CompletionUsage(
+                    prompt_tokens=count_message_tokens(messages, self.model),
```

Comment:
currently main branch has no `count_message_tokens` func. and suggest to support unify common models not only from openai.